actor/todo-sql/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+use serde::{Deserialize, Serialize};
+use wasmbus_rpc::actor::prelude::*;
+use wasmcloud_interface_httpserver::{HttpRequest, HttpResponse, HttpServer, HttpServerReceiver};
+use wasmcloud_interface_numbergen::generate_guid;
+use wasmcloud_interface_sqldb::{
+    SqlDb, SqlDbError, SqlDbSender,
+    //FetchResult, ExecuteResult,
+    minicbor,
+};
+use wasmcloud_interface_logging::{debug, info, warn};
+use once_cell::sync::Lazy;
+use minicbor::{decode,Decode};
+
+type Db = SqlDbSender<WasmHost>;
+const TABLE_NAME :&str = "todos";
+/// limit characters to avoid sql injection attacks
+/// - english letters uppercase & lowercase
+/// - digits 0-9
+/// - space, comma, period, underscore, dash, slash
+/// that's it!
+static REGEX: Lazy<regex::Regex> = Lazy::new(|| {
+    let re = regex::Regex::new(r"^[a-zA-Z0-9 ,._-/]+$").unwrap();
+    re
+});
+
+#[derive(Serialize, Deserialize)]
+struct InputTodo {
+    title: String,
+    order: Option<i32>,
+}
+#[derive(Serialize, Deserialize)]
+struct UpdateTodo {
+    title: Option<String>,
+    completed: Option<bool>,
+    order: Option<i32>,
+}
+// because 'order' is a keyword in sql, we've renamed the field 'priority',
+// but for serde serialization, leave it as 'order' so that it works
+// with existing javascript written for this app
+#[derive(Serialize, Deserialize, Decode)]
+struct Todo {
+    #[n(0)] url: String,
+    #[n(1)] title: String,
+    #[n(2)] completed: bool,
+    #[serde(rename="order")]
+    #[n(3)] priority: i32,
+}
+
+impl Todo {
+    fn new(url: String, title: String, priority: i32) -> Self {
+        Self {
+            url,
+            title,
+            completed: false,
+            priority,
+        }
+    }
+
+    fn update(self, update: UpdateTodo) -> Todo {
+        Todo {
+            url: self.url,
+            title: update.title.unwrap_or(self.title),
+            completed: update.completed.unwrap_or(self.completed),
+            priority: update.order.unwrap_or(self.priority),
+        }
+    }
+}
+
+// check safety of any user input string before it used in an sql statement
+// If the provider (&interface) supported prepared statements,
+// this could allow more characters
+fn check_safety(tag: &str, uncertain_input: &str) -> Result<(), SqlDbError> {
+    if !REGEX.is_match(&uncertain_input) {
+        return Err(SqlDbError::new("invalid",
+           format!("{} contains invalid characters", tag)))
+    }
+    Ok(())
+}
+
+/// create an empty table with the proper schema
+async fn create_table(ctx: &Context, client: &Db) -> Result<(),SqlDbError> {
+
+    let sql = format!(r#"create table {} (
```

Comment:
We should change this SQL to be idempotent... we don't want to cause panics/failures if someone hits the `init` endpoint twice.

Reply:
agreed